memlockd.8
.TH memlockd 8
.SH NAME
memlockd \- daemon to lock files in memory with mlock
.SH SYNOPSIS
\f3memlockd\f1 [ \f3\-c config-file\f1 ] [ \f3\-p pid-file\f1 ] [ \f3\-d\f1 ] [ \f3\-f\f1 ] [ \f3\-v\f1 ] [ \f3\-u user\f1 ]
.br
.SH DESCRIPTION
This manual page documents briefly the
.B memlockd
command.
.P
It is used to lock system programs and config files in memory so that if a
DOS attack is experienced then the chance of the sys\-admin regaining
control of the system in a reasonable amount of time (and therefore having
a reasonable chance of discovering the cause of the problem) is significantly
increased.
.SH OPTIONS
The \f3\-c\f1 option is used to specify the fully-qualified path name to a
config file that lists the names of files to lock, if the config file is not
specified then it will default to \f3/etc/memlockd.cfg\f1. In any situation
where a config file is used a directory can be used instead, for a directory
every file ending in ".cfg" will be processed.

The \f3\-c\f1 option is used to specify the fully-qualified path name of the
PID file that identifies a running daemon. If not specified it will default
to \f3/var/run/memlockd.pid\f1. This is useful when running multiple memlockd
daemons; for example, a stock system memlockd and an application specific
memlockd being used to prevent swapping of some application critical data
files.

The \f3\-d\f1 option specifies debugging mode, the program will not fork
and will produce it's logging messages on stderr instead of via syslog.

The \f3\-f\f1 option specifies foreground (non-daemon) mode, the program will
not fork but will still log normally.

The \f3\-v\f1 option prints program name and version.

The \f3\-u\f1 option specifies the name of a user to use for running ldd (for
recursive operation).  Note that locking shared objects that are writable by
non\-root is not safe, but using a different UID will reduce the risk.

The config file will contain a number of fully qualified names of files to
lock in RAM.  When locking shared objects and ELF binaries it is possible to
prefix the file name with a \f3+\f1 character to indicate that memlockd should
recursively lock all shared objects that the program requires and all shared
objects that those objects require. When a file not found error doesn't matter
(EG you want a single config file to have the file names for multiple
architectures or systems) you can prefix the file name with a \f3?\f1
character, in that case errors such as EPERM will still be logged.

If a line in the config file starts with a \f3%\f1 character it will be taken
as the name of a config file or directory to process.  Currently only one
level of recursion is accepted.

.TP
.SH SEE ALSO
.BR mlock (2),
.BR mmap (1).
.br
.SH AUTHOR
memlockd was written by Russell Coker <russell@coker.com.au>


debian/changelog
memlockd (1.2) unstable; urgency=low

  * allow multple deamons via cmdline option naming pidfile
  * add -v option to report version

 -- Devin Ben-Hur <dbenhur@whitepages.com>  Fri, 06 Dec 2013 12:42:12 -0800

memlockd (1.1.1) unstable; urgency=high

  * Made it work correctly with systemd, restarts via systemd and doesn't use
    a pidfile.
    Closes: #678161

 -- Russell Coker <russell@coker.com.au>  Wed, 20 Jun 2012 20:06:20 +1000

memlockd (1.1) unstable; urgency=high

  * Use wheezy paths for shared objects on i386 and amd64.  Also added a new
    config file option to not log file not found errors so we don't see i386
    errors on amd64 and amd64 errors on i386.
    Closes: #660321, #480915
  * Added systemd service file.
    Closes: #652386
  * Added a run-parts style config directory, default is /etc/memlock.d and
    the config file uses a % to chain to another file or directory.
    Closes: #529265

 -- Russell Coker <russell@coker.com.au>  Sat, 16 Jun 2012 15:43:44 +1000

memlockd (1.0) unstable; urgency=low

  * Changed the default config to have library names that work on AMD64
    and to include killall and ps.
  * Better use of variables in the makefile.
  * Put a copyright statement in the main source file and change the license
    to GPLv3.
  * Version 1.0, with no serious changes for 3 years it's good for that.

 -- Russell Coker <russell@coker.com.au>  Tue, 11 Jan 2011 21:07:34 +1100

memlockd (0.05) unstable; urgency=low

  * Stopped including asm/page.h because we didn't use it since 0.03-0.2.
    Closes: #479968

 -- Russell Coker <russell@coker.com.au>  Wed,  7 May 2008 22:06:00 +1000

memlockd (0.04-0.1) unstable; urgency=low

  * Non-maintainer upload to solve release goal.
  * Add LSB dependency header to init.d scripts (Closes: #468888).

 -- Petter Reinholdtsen <pere@debian.org>  Sat, 29 Mar 2008 14:22:03 +0100

memlockd (0.04) unstable; urgency=low

  * Include changes from Robert S. Edmonds
  * Fixed my email address in the Maintainer field.

 -- Russell Coker <russell@coker.com.au>  Sun, 27 Jan 2008 15:57:00 +1100

memlockd (0.03-0.2) unstable; urgency=low

  * Non-maintainer upload.
  * Compile with g++, not gcc -lstdc++; closes: #441449.

 -- Robert S. Edmonds <edmonds@debian.org>  Sun, 09 Sep 2007 17:17:20 -0400

memlockd (0.03-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * Use sysconf(_SC_PAGESIZE) instead of PAGE_SIZE; closes: #418666.
  * Add dependency on adduser; closes: #431529.
  * Use invoke-rc.d; closes: #440574.

 -- Robert S. Edmonds <edmonds@debian.org>  Wed, 05 Sep 2007 19:34:22 -0400

memlockd (0.03) unstable; urgency=medium

  * Fixed section and description.

  * Made it write a pidfile, and also made the restart option of the init.d
    script start a new instance unconditionally so the new instance can kill
    the old.  This means that there is no window where memlockd is not
    running.

 -- Russell Coker <russell@coker.com.au>  Tue, 13 Mar 2007 19:12:00 +1100

memlockd (0.02) unstable; urgency=medium

  * Made it not recursively call ldd.  ldd does recursion.

 -- Russell Coker <russell@coker.com.au>  Thu, 22 Feb 2007 09:25:00 +1100

memlockd (0.01) unstable; urgency=medium

  * Initial release

 -- Russell Coker <russell@coker.com.au>  Fri, 16 Feb 2007 18:19:00 +1100



memlockd.cpp
/*
 * Copyright (C) 2007-2012 Russell Coker <russell@coker.com.au>
 * Licensed under GPL v3
 */
#include <assert.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <syslog.h>
#include <signal.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <stdarg.h>
#include <dirent.h>

#define MAX_FILES 1024
#define PIDFILE "/var/run/memlockd.pid"

typedef struct file_data
{
  char *name;
  int fd;
  struct stat sb;
  void *start;
  int map_size;
} FILE_DATA;

FILE_DATA files[MAX_FILES];
FILE_DATA new_files[MAX_FILES];
int num_files = 0;
int num_new_files = 0;
const char * config = "/etc/memlockd.cfg";
int debug = 0;
int page_size = 0;
uid_t uid = 0;
gid_t gid = 0;

#define BUF_SIZE 1024

void log(int priority, const char * const format, ...)
{
  va_list argp;
  va_start(argp, format);
  if(debug)
  {
    vfprintf(stderr, format, argp);
    fprintf(stderr, "\n");
  }
  else
    vsyslog(priority, format, argp);
}

void unmap_file(FILE_DATA *data)
{
  munmap(data->start, data->sb.st_size);
  log(LOG_INFO, "Unmapped file %s", data->name);
}

void unmap_close_file(FILE_DATA *data)
{
  unmap_file(data);
  free(data->name);
  data->name = NULL;
  close(data->fd);
  data->fd = -1;
}

// return 1 if a file is mapped
int open_map(int fd, struct stat *sb, const char * const name)
{
  new_files[num_new_files].start = mmap(NULL, sb->st_size, PROT_READ, MAP_SHARED, fd, 0);
  if(new_files[num_new_files].start == MAP_FAILED)
  {
    log(LOG_ERR, "Error mmaping %s: %s", name, strerror(errno));
    close(fd);
    return 0;
  }
  if(mlock(new_files[num_new_files].start, sb->st_size) == -1)
  {
    log(LOG_ERR, "Can't lock memory for %s, error %s", name, strerror(errno));
    munmap(new_files[num_new_files].start, sb->st_size);
    close(fd);
    return 0;
  }
  if(sb->st_size % page_size)
    new_files[num_new_files].map_size = sb->st_size - (sb->st_size % page_size)
                                      + page_size;
  else
    new_files[num_new_files].map_size = sb->st_size;
  new_files[num_new_files].fd = fd;
  memcpy(&new_files[num_new_files].sb, sb, sizeof(struct stat));
  new_files[num_new_files].name = strdup(name);
  num_new_files++;
  log(LOG_INFO, "Mapped file %s", name);
  return 1;
}

// return 0 for no file mapped and 1 for file mapped
int open_file(const char * const name, int no_error_non_exist)
{
  int fd = open(name, O_RDONLY);
  if(fd == -1 && no_error_non_exist && errno == ENOENT)
    return 0;
  if(fd == -1)
  {
    log(LOG_ERR, "Can't open file %s", name);
    return 0;
  }
  struct stat sb;
  if(fstat(fd, &sb) == -1)
  {
    log(LOG_ERR, "Can't stat file %s", name);
    close(fd);
    return 0;
  }
  int i;
  for(i = 0; i < num_files; i++)
  {
    if(files[i].fd != -1 && files[i].sb.st_dev == sb.st_dev
      && files[i].sb.st_ino == sb.st_ino)
    {
      if(files[i].sb.st_size == sb.st_size
       && files[i].sb.st_mtime == sb.st_mtime)
      {
        memcpy(&new_files[num_new_files], &files[i], sizeof(FILE_DATA));
        files[i].fd = -1;
        files[i].name = NULL;
        num_new_files++;
        return 1;
      }
      else
      {
        memcpy(&new_files[num_new_files], &files[i], sizeof(FILE_DATA));
        close(fd);
        num_new_files++;
        files[i].fd = -1;
        files[i].name = NULL;
        unmap_file(&new_files[num_new_files - 1]);
        open_map(new_files[num_new_files - 1].fd
               , &new_files[num_new_files - 1].sb, name);
        return 1;
      }
    }
  }
  for(i = 0; i < num_new_files; i++)
  {
    if(new_files[i].fd != -1 && new_files[i].sb.st_ino == sb.st_ino)
    {
      close(fd);
      return 0;
    }
  }
  return open_map(fd, &sb, name);
}

void map_file_dependencies(const char * const name, int no_error_non_exist)
{
  if(!uid || !gid)
    return;
  int pipe_fd[2];
  
  if(pipe(pipe_fd) == -1)
  {
    log(LOG_ERR, "Can't create pipe, not recursing");
    uid = 0;
    return;
  }
  int rc = fork();
  if(rc == -1)
  {
    log(LOG_ERR, "Can't fork, not recursing");
    uid = 0;
    return;
  }
  if(!rc)
  {
    char buf[4096];

    close(pipe_fd[0]);
    close(1);
    if(dup2(pipe_fd[1], 1) == -1)
    {
      log(LOG_ERR, "Can't create pipe");
      exit(1);
    }
    if(setresgid(gid, gid, gid) == -1 || setresuid(uid, uid, uid) == -1)
    {
      log(LOG_ERR, "Can't set UID and GID");
      exit(1);
    }
    sprintf(buf, "/usr/bin/ldd %s", name);
    char *argv[3];
    argv[0] = strdup("/usr/bin/ldd");
    argv[1] = strdup(name);
    argv[2] = NULL;
    execv(argv[0], (char * const *)argv);
    log(LOG_ERR, "Can't exec ldd");
    exit(1);
  }
  close(pipe_fd[1]);
  FILE *fp = fdopen(pipe_fd[0], "r");
  if(!fp)
    return;

  char buf[4096];
  while(fgets(buf, sizeof(buf), fp))
  {
    char *tmp = strchr(buf, '/');
    if(!tmp)
      continue;
    strtok(tmp, " ");
    open_file(tmp, no_error_non_exist);
  }
  fclose(fp);
  wait(&rc);
}

void parse_config_file(const char * const config_name, int recurse_count)
{
  struct stat sbuf;
  if(stat(config_name, &sbuf))
  {
    log(LOG_ERR, "Can't stat \"%s\"", config_name);
    exit(1);
  }
  if(S_ISDIR(sbuf.st_mode))
  {
    log(LOG_INFO, "Entering config dir \"%s\"", config_name);
    DIR *dirp = opendir(config_name);
    if(!dirp)
    {
      log(LOG_ERR, "Can't open config file/dir %s", config_name);
      exit(1);
    }
    int rc;
    struct dirent entry, *res;
    while((rc = readdir_r(dirp, &entry, &res)) == 0 && res)
    {
      int len = strlen(entry.d_name);
      if(len > 4 && !strcmp(".cfg", &entry.d_name[len - 4]))
      {
        char buf[1024];
        snprintf(buf, sizeof(buf), "%s/%s", config_name, entry.d_name);
        parse_config_file(buf, recurse_count);
      }
    }
    if(rc)
    {
      log(LOG_ERR, "readdir_r() error for \"%s\"", config_name);
      exit(1);
    }
    closedir(dirp);
    return;
  }
  log(LOG_INFO, "Parsing config file \"%s\"", config_name);
  FILE *fp = fopen(config_name, "r");
  char buf[BUF_SIZE];
  while(fgets(buf, BUF_SIZE, fp))
  {
    int len = strlen(buf) - 1;
    if(buf[0] == '#')
      continue;
    if(buf[len] == '\n')
      buf[len] = 0;
    const char *ptr = buf;
    int map_dependencies = 0, no_error_non_exist = 0;
    if(*ptr == '%')
    {
      if(recurse_count > 1)
      {
        log(LOG_ERR, "Too much recursion, won't process \"%s\"", ptr+1);
      }
      else
      {
        ptr++;
        log(LOG_INFO, "Recursion, entering \"%s\"", ptr);
        parse_config_file(ptr, recurse_count + 1);
        continue;
      }
    }
    if(*ptr == '?')
    {
       ptr++;
       no_error_non_exist = 1;
    }
    if(*ptr == '+')
    {
       ptr++;
       map_dependencies = 1;
    }
    if(*ptr == '?')
    {
       ptr++;
       no_error_non_exist = 1;
    }
    if(*ptr != '/')
      continue;
    open_file(ptr, no_error_non_exist);
    if(map_dependencies)
      map_file_dependencies(ptr, no_error_non_exist);
  }
  fclose(fp);
}

void parse_config(int)
{
  num_new_files = 0;
  parse_config_file(config, 0);
  for(int i = 0; i < num_files; i++)
  {
    if(files[i].fd != -1)
      unmap_close_file(&files[i]);
  }
  if(!num_new_files)
  {
    log(LOG_INFO, "No files to lock - exiting");
    exit(0);
  }
  memcpy(files, new_files, sizeof(FILE_DATA) * num_new_files);
  num_files = num_new_files;
  num_new_files = 0;
}

void usage()
{
  fprintf(stderr, "Usage: memlockd [-c config-file] [-d] [-f]\n"
                  "       -d is for debugging mode (running in foreground and no syslog)\n"
                  "       -f is for foreground mode with syslog logging\n");
  exit(1);
}
int main(int argc, char **argv)
{
  int c, foreground = 0;
  pid_t old_pid = 0;
  page_size = (int) sysconf(_SC_PAGESIZE);
  while(-1 != (c = getopt(argc, argv, "fdc:u:")) )
  {
    switch(char(c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'c':
        config = optarg;
      break;
      case 'd':
        debug = 1;
      break;
      case 'f':
        foreground = 1;
      break;
      case 'u':
        struct passwd *pw = getpwnam(optarg);
        if(!pw)
        {
          log(LOG_ERR, "Can't look up user %s", optarg);
          exit(1);
        }
        uid = pw->pw_uid;
        gid = pw->pw_gid;
        endpwent();
      break;
    }
  }

  openlog("memlockd", LOG_CONS | LOG_PID, LOG_DAEMON);

  int write_pidfile = 1;
  if(debug || foreground || getuid())
    write_pidfile = 0;

  if(!debug && !foreground)
    daemon(0, 0);
  else
    chdir("/");
  if(write_pidfile)
  {
    FILE *fp = fopen(PIDFILE, "r");
    char buf[20];
    if(fp)
    {
      if(fgets(buf, sizeof(buf), fp))
        old_pid = atoi(buf);
      else
        log(LOG_ERR, "Can't read pidfile " PIDFILE);
      fclose(fp);
    }
  }

  if(mlockall(MCL_CURRENT|MCL_FUTURE) == -1)
  {
    log(LOG_ERR, "Can't lock memory, exiting");
    exit(1);
  }
  parse_config(0);

  struct sigaction sa;
  sa.sa_sigaction = NULL;
  sa.sa_flags = SA_SIGINFO;
  sa.sa_handler = parse_config;
  if(sigaction(SIGHUP, &sa, NULL))
    log(LOG_ERR, "Can't handle sighup");
  if(write_pidfile)
  {
    FILE *fp = fopen(PIDFILE, "w");
    if(fp)
    {
      if(fprintf(fp, "%d", (int)getpid()) <= 0)
      {
        log(LOG_ERR, "Can't write to " PIDFILE);
        unlink(PIDFILE);
      }
      fclose(fp);
    }
    else
      log(LOG_ERR, "Can't open " PIDFILE " for writing");
  }
  if(old_pid)
    kill(old_pid, SIGKILL);
  while(1)
    sleep(3600);
  return 0;
}
